sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeCreator.scala

```diff
@@ -18,14 +18,71 @@ package org.apache.spark.sql.rapids
 
 import ai.rapids.cudf.ColumnVector
 import com.nvidia.spark.rapids.{GpuColumnVector, GpuExpression, GpuScalar}
-import com.nvidia.spark.rapids.RapidsPluginImplicits.{AutoCloseableArray, ReallyAGpuExpression}
+import com.nvidia.spark.rapids.RapidsPluginImplicits.ReallyAGpuExpression
 
+import org.apache.spark.sql.catalyst.analysis.{TypeCheckResult, TypeCoercion}
 import org.apache.spark.sql.catalyst.analysis.FunctionRegistry.FUNC_ALIAS
-import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
 import org.apache.spark.sql.catalyst.expressions.{EmptyRow, Expression, NamedExpression}
-import org.apache.spark.sql.types.{Metadata, StringType, StructField, StructType}
+import org.apache.spark.sql.catalyst.util.TypeUtils
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{ArrayType, DataType, Metadata, NullType, StringType, StructField, StructType}
 import org.apache.spark.sql.vectorized.ColumnarBatch
 
+case class GpuCreateArray(children: Seq[Expression], useStringTypeWhenEmpty: Boolean)
+    extends GpuExpression {
+
+  def this(children: Seq[Expression]) = {
+    this(children, SQLConf.get.getConf(SQLConf.LEGACY_CREATE_EMPTY_COLLECTION_USING_STRING_TYPE))
+  }
+
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def stringArgs: Iterator[Any] = super.stringArgs.take(1)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    TypeUtils.checkForSameTypeInputExpr(children.map(_.dataType), s"function $prettyName")
+  }
+
+  private val defaultElementType: DataType = {
+    if (useStringTypeWhenEmpty) {
+      StringType
+    } else {
+      NullType
+    }
+  }
+
+  override def dataType: ArrayType = {
+    ArrayType(
+      TypeCoercion.findCommonTypeDifferentOnlyInNullFlags(children.map(_.dataType))
+          .getOrElse(defaultElementType),
+      containsNull = children.exists(_.nullable))
+  }
+
+  override def nullable: Boolean = false
+
+  override def prettyName: String = "array"
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    withResource(new Array[ColumnVector](children.size)) { columns =>
+      val numRows = batch.numRows()
+      children.indices.foreach { index =>
+        children(index).columnarEval(batch) match {
+          case cv: GpuColumnVector =>
+            columns(index) = cv.getBase
+          case other =>
+            val dt = dataType.elementType
+            withResource(GpuScalar.from(other, dt)) { scalar =>
+              columns(index) = ColumnVector.fromScalar(scalar, numRows)
```

Comment:
Should there be a TODO comment and/or followup issue to avoid manifesting splatted scalar columns for each list entry when libcudf supports nested scalar types which could be splatted directly?

Reply:
Spark will already take care of this for us.  `array(1, 2, 3)` is already folded into a scalar array, which then we have to fall back to the CPU for because we cannot splat it.  This code is only for the mixed case where we have a mixture of columns and scalar values.  Unfortunately the underlying CUDF API also does not support scalars so we could not even save memory and avoid splatting the individual scalar values, not even in a detailed API.  I am happy to file a follow on issue if that is something that we feel would be good to try and support.

Reply:
> array(1, 2, 3) is already folded into a scalar array, which then we have to fall back to the CPU for because we cannot splat it.

Ouch, that's unfortunate.  Even more cause to seek nested type scalar support in libcudf.

> I am happy to file a follow on issue if that is something that we feel would be good to try and support.

It's probably not worth it yet.  We're already splatting scalars into columns everywhere else.  Let's see if that ends up being an issue in practice, and we already have a separate issue to track the general splatting of scalars to columns in expressions.